src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+                        pSrcCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static void MulElementWiseU(float[] src1, float[] src2, float[] dst, int count)
+        {
+            unsafe
+            {
+                fixed (float* psrc1 = &src1[0])
+                fixed (float* psrc2 = &src2[0])
+                fixed (float* pdst = dst)
+                {
+                    float* pSrc1Current = psrc1;
+                    float* pSrc2Current = psrc2;
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> src1Vector = Sse.LoadVector128(pSrc1Current);
+                        Vector128<float> src2Vector = Sse.LoadVector128(pSrc2Current);
+                        src2Vector = Sse.Multiply(src1Vector, src2Vector);
+                        Sse.Store(pDstCurrent, src2Vector);
+
+                        pSrc1Current += 4;
+                        pSrc2Current += 4;
+                        pDstCurrent += 4;
+                    }
+
+                    while (pDstCurrent < pEnd)
+                    {
+                        Vector128<float> src1Vector = Sse.LoadScalarVector128(pSrc1Current);
+                        Vector128<float> src2Vector = Sse.LoadScalarVector128(pSrc2Current);
+                        src2Vector = Sse.MultiplyScalar(src1Vector, src2Vector);
+                        Sse.StoreScalar(pDstCurrent, src2Vector);
+
+                        pSrc1Current++;
+                        pSrc2Current++;
+                        pDstCurrent++;
+                    }
+                }
+            }
+        }
+
+        internal static float SumSqU(float[] src, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pEnd = psrc + count;
+
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                        result = Sse.Add(result, Sse.Multiply(srcVector, srcVector));
+
+                        pSrcCurrent += 4;
+                    }
+
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
+                        result = Sse3.HorizontalAdd(result, result);
+                        result = Sse3.HorizontalAdd(result, result);
+                    }
+                    else
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
```

Comment:
The second `Sse.MoveHighToLow(result, Sse.UnpackLow(result, result))` is probably more efficient as: `Shuffle(result, result, (1, 0, 3, 2))`